### PR TITLE
Added Yii 3 DI compatibility

### PR DESCRIPTION
**Pull request by hiqsol**
- Allowed `__class` and `__construct()` in configurations:
    - `'__class' => SomeClass::class`
    - `'__construct()' => ['ConstructorArg1', 'ConstructorArg2']`
- Allowed wider use of `Instance::of()`:
    - in definitions: `'SomeInterface' => Instance::of('SomeService')`
    - in get: `$container->get(Instance::of('SomeInterface')`
- Allowed static call DI definition:
    - `'SomeInterface' => [SomeFactory::class, 'createMethod']`
- Added more tests

| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Tests pass?   | ✔️

### CODE REVIEW

**Comment by samdark** [comment]
Seems it doesn't work with PHP < 7.1: https://travis-ci.org/yiisoft/yii2/jobs/597063005

File: tests/framework/di/ContainerTest.php [hunk 1]
```diff
@@ -282,6 +282,66 @@ public function testSetDependencies()
...
+            ],
+        ]);
+        $bar = $container->get('bar');
+        $this->assertInstanceOf(Bar::className(), $bar);
+        $qux = $bar->qux;
+        $this->assertInstanceOf(Qux::className(), $qux);
+        $this->assertSame(42, $qux->a);
+    }
+
+    public function testGetByInstance(): void
```
Comment (rob006) [inline tests/framework/di/ContainerTest.php]:
```suggestion
    public function testGetByInstance()
```

**Comment by samdark** [comment]
Documentation needs an update as well.

**Comment by samdark** [comment]
@hiqsol do you have time for taking care of docs?

**Comment by hiqsol** [comment]
> @hiqsol do you have time for taking care of docs?

This guide?
https://github.com/yiisoft/yii2/blob/master/docs/guide/concept-di-container.md

Other places?

**Comment by samdark** [comment]
Guide, yes.

**Review by samdark — APPROVED** [review]

**Pull request merged**